File: c_programs/read_weight1.py
#!/usr/bin/env python3
import time
import sys

# Define GPIO pins for HX711
GPIO_DOUT = 69
GPIO_SCK = 68

# --- Sysfs GPIO helper functions (no changes here) ---
def gpio_export(pin):
    try:
        with open("/sys/class/gpio/export", "w") as f:
            f.write(str(pin))
    except (IOError, PermissionError):
        # Already exported, or permission issue
        # print(f"Warning: Could not export pin {pin}. Already exported?")
        pass

def gpio_unexport(pin):
    try:
        with open("/sys/class/gpio/unexport", "w") as f:
            f.write(str(pin))
    except (IOError, PermissionError):
        print(f"Warning: Could not unexport pin {pin}.")

def gpio_set_dir(pin, direction):
    with open(f"/sys/class/gpio/gpio{pin}/direction", "w") as f:
        f.write(direction)

def gpio_set_value(pin, value):
    with open(f"/sys/class/gpio/gpio{pin}/value", "w") as f:
        f.write("1" if value else "0")

def gpio_get_value(pin):
    with open(f"/sys/class/gpio/gpio{pin}/value", "r") as f:
        return int(f.read().strip())

# --- Setup and Cleanup ---
def setup():
    print("Setting up GPIO pins...")
    gpio_export(GPIO_DOUT)
    gpio_export(GPIO_SCK)
    # Allow a moment for sysfs to create the directories
    time.sleep(0.1)
    gpio_set_dir(GPIO_DOUT, "in")
    gpio_set_dir(GPIO_SCK, "out")
    gpio_set_value(GPIO_SCK, 0)
    print("Setup complete.")

def cleanup():
    print("\nCleaning up GPIO...")
    gpio_unexport(GPIO_DOUT)
    gpio_unexport(GPIO_SCK)
    print("Cleanup complete.")

# --- MODIFIED HX711 Read Function ---
def hx711_read():
    """
    Reads a 24-bit value from the HX711.
    Returns the raw integer value or None on timeout.
    """
    # 1. Wait for the DOUT pin to go LOW
    ready_timeout = time.time() + 0.5  # 500ms timeout
    while gpio_get_value(GPIO_DOUT) == 1:
        if time.time() > ready_timeout:
            print("Error: Timeout waiting for HX711 to become ready.")
            return None
    
    # 2. Read the 24 bits of data
    count = 0
    # The delays from sysfs I/O are often sufficient,
    # so we remove the explicit time.sleep() calls here.
    for _ in range(24):
        gpio_set_value(GPIO_SCK, 1)
        count = count << 1
        gpio_set_value(GPIO_SCK, 0)
        if gpio_get_value(GPIO_DOUT):
            count += 1

    # 3. Send one more pulse to set gain for the next reading
    gpio_set_value(GPIO_SCK, 1)
    gpio_set_value(GPIO_SCK, 0)

    # 4. Check if the value is negative (using 2's complement)
    if (count & 0x800000):
        count |= ~0xffffff

    return count

# --- Main Program Execution ---
if __name__ == "__main__":
    try:
        setup()
        # Give the sensor time to settle
        time.sleep(1)
        
        while True:
            val = hx711_read()
            if val is not None:
                print(f"Raw weight data: {val}")
            else:
                # If a timeout occurred, wait before retrying
                print("Retrying after timeout...")
            time.sleep(0.5)

    except KeyboardInterrupt:
        print("Exiting program.")
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        cleanup()

File: c_programs/read_weight1.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>

#define GPIO_DOUT 69 // GPIO pin for HX711 DOUT
#define GPIO_SCK  68 // GPIO pin for HX711 SCK

// Export GPIO pin in sysfs
void gpio_export(int gpio) {
    char buffer[64];
    int len = snprintf(buffer, sizeof(buffer), "%d", gpio);
    int fd = open("/sys/class/gpio/export", O_WRONLY);
    write(fd, buffer, len);
    close(fd);
}

// Set GPIO direction
void gpio_set_dir(int gpio, const char* dir) {
    char path[64];
    snprintf(path, sizeof(path), "/sys/class/gpio/gpio%d/direction", gpio);
    int fd = open(path, O_WRONLY);
    write(fd, dir, strlen(dir));
    close(fd);
}

// Write GPIO value
void gpio_set_value(int gpio, int value) {
    char path[64];
    snprintf(path, sizeof(path), "/sys/class/gpio/gpio%d/value", gpio);
    int fd = open(path, O_WRONLY);
    if(value)
        write(fd, "1", 1);
    else
        write(fd, "0", 1);
    close(fd);
}

// Read GPIO value
int gpio_get_value(int gpio) {
    char path[64], value_str[3];
    snprintf(path, sizeof(path), "/sys/class/gpio/gpio%d/value", gpio);
    int fd = open(path, O_RDONLY);
    read(fd, value_str, 3);
    close(fd);
    return atoi(value_str);
}

// HX711 read 24-bit value with configurable gain pulses
long hx711_read() {
    long count;
    unsigned char i;
    int gain_pulses = 3; // 1 for gain 128, 2 for gain 32, 3 for gain 64

    // Wait for DOUT to go low (data ready)
    while(gpio_get_value(GPIO_DOUT));

    count = 0;
    for(i = 0; i < 24; i++) {
        gpio_set_value(GPIO_SCK, 1);
        usleep(100);   // increased delay for stable reading
        count = count << 1;
        gpio_set_value(GPIO_SCK, 0);
        int bit = gpio_get_value(GPIO_DOUT);
        if(bit)
            count++;
        printf("Bit %d: %d\n", i, bit);  // debug print each bit
        usleep(100);
    }

    // Send clock pulses to set gain/channel
    for(int j = 0; j < gain_pulses; j++) {
        gpio_set_value(GPIO_SCK, 1);
        usleep(100);
        gpio_set_value(GPIO_SCK, 0);
        usleep(100);
    }

    // convert 24-bit to signed long
    if(count & 0x800000)
        count |= ~0xffffff;

    return count;
}

int main() {
    // Initialize GPIO pins
    gpio_export(GPIO_DOUT);
    gpio_export(GPIO_SCK);
    gpio_set_dir(GPIO_DOUT, "in");
    gpio_set_dir(GPIO_SCK, "out");

    // Set initial SCK low
    gpio_set_value(GPIO_SCK, 0);

    while(1) {
        long weight_raw = hx711_read();
        printf("Raw weight data: %ld\n", weight_raw);
        usleep(500000); // 0.5s delay between readings
    }

    return 0;
}
